### ISSUE TITLE
[Request] Extract MP4-Metadata with ffprobe

### ISSUE THREAD

**Issue by viaregio** [jellyfin/jellyfin#2407]
I have a lot of MP4-videos with embedded metadata (title, description, season number, episode number, ...). Until now I managed the files with iTunes, so the filenames and folder structure is managed by iTunes. The titles are mainly in two languages: German and English. My setup of Jellyfin consists of four Libraries: Movies, TV-Shows, Music and Home videos. The problem Jellyfin completely ignores the embedded metadata from iTunes and solely tries to get season and episode numbers from the file path.
 
**Expected behavior**
It would be nice if Jellyfin would read the embedded metadata of the MP4-files and use this metadata to fetch further informations based on this metadata

**Additional context**
I think the part of the code that needs to be changed is the EpisodeResolver in Emby.Naming.TV. The code for reading the metadata already exists in ` jellyfin/MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs `.
ffprobe can be called with:
```ffprobe -v quiet -i filename.mp4 -of json -show_format```
The resulting json contains a `format` key and the data under `format` contains a `tags` key. Under the `tags` key the interesting keys are:

  * title
  * description
  * synopsis
  * media_type (0 for home videos, 9 for movies, 10 for tv shows)
  * show
  * season_number
  * episode_sort
  * network

**Comment by Artiume** [comment]
I think the proper way to handle enabling a metadata crawler is to first allow it to be enabled or disabled. If I had this enabled on my library, it could have effects I wouldn't desire.

On that note, we should be able to give preference for external, nfo and metadata order. 

What containers can hold metadata? What tools will be required to extract metadata? Can ffprobe extract all important metadata?

**Comment by JustAMan** [comment]
Note that we track feature requests separately: https://features.jellyfin.org/
I'm not closing this one yet as _maybe_ it could be a simple change.

**Comment by viaregio** [comment]
I think I could deliver the code that is extracting the metadata from a given file path in a pull request but I'm very new to the code of Jellyfin. Is ` jellyfin/Emby.Naming/TV/EpisodeResolver.cs ` really the right place to implement this code? I have no idea where to add the code for the order of the metadata sources and I think @Artiume has a very good point in this should be optional.

If someone can give me a hint where to look, I will try to implement this configuration options also.

**Comment by Artiume** [comment]
https://github.com/jellyfin/jellyfin/blob/master/Emby.Naming/TV/EpisodeResolver.cs

This looks more like where the metadata should be retained. I don't see any scanner logic here.

**Comment by viaregio** [comment]
In `jellyfin/Emby.Naming/TV/EpisodePathParser.cs` season and episode numbers are extracted from the filename. The result is saved in `parsingResult` and for the mp4-container I would, if enabled, overwrite SeasonNumber, EpisodeNumber and SeriesName with the metadata from ffprobe.

**Comment by Artiume** [comment]
Yeah, so the issue is we need to find the global controller, inject the boolean that'll turn embedded scanner off and on (off for performance by default) and so we can build the preferences of nfo > internet > metadata and so on.

This file is specifically isolated to your issue, tv shows. But we need to make it available for movies and so on. I want to say we need to look in MediaBrowser.Api

**Comment by dkanada** [comment]
I would start by looking for the logic that pulls tags from FLAC files (I can confirm this works) and then implement the same method for MP3 files if it doesn't exist.

**Comment by Artiume** [comment]
So mp3 id3 tags are always scraped but I'm not sure if they're kept. For OP's original issue, it's specifically for MP4 videos

**Comment by viaregio** [comment]
If I analyzed the code correctly, the fetching of the metadata for TV shows is a two staged process:

  1. Episode and season numbers are determined by analyzing the path of the video file
  2. After episode and season numbers are determined further metadata is fetched with the configured Providers

I made two pull requests for `jellyfin` and `jellfin-web` to solve a problem with stage 1. If the episode number can't be determined correctly by the path of the video file there is little change to fetch the further metadata correctly.
